### PR TITLE
extract controller name logic into another method

### PR DESCRIPTION
**Pull request by steinkel**

### CODE REVIEW

File: src/Http/ControllerFactory.php [hunk 1]
```diff
@@ -65,16 +91,7 @@ public function create(ServerRequest $request, Response $response)
...
-        if (!$className) {
-            $this->missingController($request);
-        }
-        $reflection = new ReflectionClass($className);
-        if ($reflection->isAbstract() || $reflection->isInterface()) {
-            $this->missingController($request);
-        }
-
-        return $reflection->newInstance($request, $response, $controller);
+        return App::className($pluginPath . $controller, $namespace, 'Controller');
```
Comment (dereuromark) [inline src/Http/ControllerFactory.php]: return ... ?: null;

would be cleaner, returning more than null as alternative is a code smell usually.

File: src/Http/ControllerFactory.php [hunk 2]
```diff
@@ -33,14 +33,40 @@ class ControllerFactory
...
+        }
+
+        return $reflection->newInstance($request, $response, $controller);
+    }
+
+    /**
+     * Determine the controller class name based on current request and controller param
+     *
+     * @param \Cake\Http\ServerRequest $request The request to build a controller for.
+     * @param string $controllerName The controller name present in the request params
```
Comment (dereuromark) [inline src/Http/ControllerFactory.php]: |null missing

**Comment by steinkel** [comment]
@dereuromark thank you for the review

**Comment by dereuromark** [comment]
This looks BC, could also have gone in master. But 3.next sure is also fine.

File: src/Http/ControllerFactory.php [hunk 3]
```diff
@@ -33,14 +33,40 @@ class ControllerFactory
      */
     public function create(ServerRequest $request, Response $response)
     {
-        $pluginPath = $controller = null;
+        $controller = null;
+        if ($request->getParam('controller')) {
+            $controller = $request->getParam('controller');
+        }
```
Comment (markstory) [inline src/Http/ControllerFactory.php]: Do you need these 3 lines? They are duplicated on 64-66.

**Review by markstory — COMMENTED** [review]
I'm guessing you have a use case for this being a separate method? What is it?

File: src/Http/ControllerFactory.php (hunk 3 above)
Comment (steinkel) [inline src/Http/ControllerFactory.php]: Trying to keep it simple as $controller is used in below https://github.com/cakephp/cakephp/pull/10558/files/4a49f1e8393bb259cf36c4455d1b42f87bdc4e63#diff-4c9ecd90d3740d54e09028ff7be12e1eR49

But makes sense to allow others to use the new getControllerClass method without passing the controller.

**Comment by steinkel** [comment]
The use case for this is: while working on this ticket https://github.com/CakeDC/users/issues/527 I got to a point where I would like to check if the controller is exactly the same as the controller specified by a set of params (a rbac rule). Following the the dispatch process I found this code as part of the controller factory, so the refactor.

File: src/Http/ControllerFactory.php (hunk 3 above)
Comment (markstory) [inline src/Http/ControllerFactory.php]: Then could the parameter be required for `getControllerClass`?

Comment (steinkel) [inline src/Http/ControllerFactory.php]: I added as optional to let others pass only the request.

Comment (markstory) [inline src/Http/ControllerFactory.php]: I would like to not have the duplicate code though.

Comment (steinkel) [inline src/Http/ControllerFactory.php]: ok, I'll take a look later to refactor that too, thanks

File: src/Http/ControllerFactory.php [hunk 4]
```diff
@@ -46,22 +42,20 @@ public function create(ServerRequest $request, Response $response)
             $this->missingController($request);
         }
 
-        return $reflection->newInstance($request, $response, $controller);
+        return $reflection->newInstance($request, $response);
```
Comment (markstory) [inline src/Http/ControllerFactory.php]: This should be safe. The controller uses the request params as a fallback for the name.

File: tests/TestCase/Http/ControllerFactoryTest.php [hunk 5]
```diff
@@ -263,4 +263,23 @@ public function testAbsoluteReferenceFailure()
         ]);
         $this->factory->create($request, $this->response);
     }
+
+    /**
+     * Test building controller name passing no controller name
```
Comment (ravage84) [inline tests/TestCase/Http/ControllerFactoryTest.php]: Sentence is somewhat confusing, may be changing to "name when passing" helps )If I understood correctly).

Comment (steinkel) [inline tests/TestCase/Http/ControllerFactoryTest.php]: fixed, thanks

**Pull request merged**